lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -411,178 +444,152 @@ utxoFromUnvalidatedTx :: Tx -> UTxO
-    :: IsOurs s RewardAccount
-    => DelegationCertificate
-    -> State s (Maybe DelegationCertificate)
-ourDelegation cert =
-    state (isOurs $ dlgCertAccount cert) <&> \case
-        Nothing -> Nothing
-        Just{}  -> Just cert
-
-ourWithdrawal
-    :: IsOurs s RewardAccount
-    => (RewardAccount, Coin)
-    -> State s (Maybe (RewardAccount, Coin))
-ourWithdrawal (acct, amt) =
-    state (isOurs acct) <&> \case
-        Nothing -> Nothing
-        Just{}  -> Just (acct, amt)
-
-ourWithdrawalSumFromTx
-    :: IsOurs s RewardAccount
-    => Tx
-    -> State s Coin
-ourWithdrawalSumFromTx tx
-    -- If a transaction has failed script validation, then the ledger rules
-    -- require that applying the transaction shall have no effect other than
-    -- to fully spend the collateral inputs included within that transaction.
-    --
-    -- Therefore, any reward withdrawals included in such a transaction should
-    -- also have no effect.
-    --
-    | failedScriptValidation tx =
-        pure (Coin 0)
-    | otherwise =
-        F.foldMap snd <$>
-        mapMaybeM ourWithdrawal (Map.toList $ withdrawals tx)
+{-------------------------------------------------------------------------------
+                        Address ownership and discovery
+-------------------------------------------------------------------------------}
 
--- | Indicates whether a given transaction is relevant to the wallet.
---
-isOurTx
-    :: forall s. (IsOurs s Address, IsOurs s RewardAccount)
-    => Tx
-    -> UTxO
-    -> State s Bool
-isOurTx tx u
-    -- If a transaction has failed script validation, then the ledger rules
-    -- require that applying the transaction shall have no effect other than
-    -- to fully spend the collateral inputs included within that transaction.
-    --
-    -- Therefore, such a transaction is only relevant to the wallet if it has
-    -- one more collateral inputs that belong to the wallet.
-    --
-    | failedScriptValidation tx =
-        txHasRelevantCollateral
-    | otherwise =
-        F.or <$> sequence
-            [ txHasRelevantInput
-            , txHasRelevantOutput
-            , txHasRelevantWithdrawal
-            ]
+-- | Perform address discovery by going through all transactions
+-- and delegation certificates in the block.
+discoverAddresses
+    :: (IsOurs s Address, IsOurs s RewardAccount)
+    => Block -> s -> s
+discoverAddresses block s0 = s2
   where
-    txHasRelevantCollateral =
-        pure . not . UTxO.null $
-        u `UTxO.restrictedBy` Set.fromList (fst <$> tx ^. #resolvedCollateral)
-    txHasRelevantInput =
-        pure . not . UTxO.null $
-        u `UTxO.restrictedBy` Set.fromList (fst <$> tx ^. #resolvedInputs)
-    txHasRelevantOutput =
-        F.or <$> sequence (isOurAddress . (^. #address) <$> tx ^. #outputs)
-    txHasRelevantWithdrawal =
-        F.or <$> sequence (isOurWithdrawal <$> Map.toList (tx ^. #withdrawals))
+    -- NOTE: Order in which we perform discovery is important.
```

Comment:
We not only perform operations in sequence, but also need to collect the results with `F.or`. In order to collect results with `do`, we would need to bring new identifiers into scope, but they would be boilerplate. The present form is more succinct.

Anyway, since `isOurTx` is used for testing only, I would leave it as it is.